Create sidebar

DylanMeeus opened the pull request:
Created a sidebar and associated components (sidebarservice, models, builders). The style on the sidebar might have to change, but it has some css applied to it already. 

I have also created a "/Logout" entry in the routeparams so we can use a link to log out a user. And a simple 404 page for routes that do not yet exist, such as the "edit" links. 

this PR connects to #12

tdonohue commented on resources/i18n/en.json:
```diff
@@ -125,5 +125,87 @@
+    
+    "item.create.processing": "Item {{name}} is being created. Please wait.",
+    "item.create.success": "Item {{name}} was created successfully.",
+    "item.create.error": "Error creating item {{name}}.",
```
@DylanMeeus : this seems to accidentally duplicate all the Item metadata/create language tags in `en.json`.  This tags were all moved up recently to reorder the `en.json` alphabetically.

tdonohue commented:
I've tested this and it works. I like how the sidebar is modular (e.g. page specific sidebars build on the app sidebar). Very cool!

This could use some more code comments. There's a lot of empty code comments (i.e. methods that have no explanation, but may just list their params...not all of these are self-explanatory). Admittedly, I'm still trying to learn how some of this is working (especially the observable stuff), so it's at least helpful to me to have code comments.

Overall, I'm not against merging this. It's good work. I'd just like to see more code comments.  Oh, and as noted above, `en.json` now has some duplicative keys (very minor issue).

DylanMeeus commented:
@tdonohue: Thank you for the feedback, I have removed the duplicate keys in the en.json file and I have added more comments. The non-self-explanatory methods (and variables) now have comments explaining why they are there and what their parameters are for.

ghost commented:
I like the work done but have a few considerations. I see a few components sharing properties and methods that should probably be an extension of a class that implements an interface. I also would like to see scope level providers for dependency injection over the instantiation done within many of the new components. As for the context of given sidebar components, they should receive its context through the context provider, which should be injected. 

These are some things I think should be considered but do not feel are a requirement of this pull request. This is good work and provides the initial groundwork for the sidebar. As I can tell, looks good to merge. I would propose additional work be done to make improvements that are of consensus.

DylanMeeus commented:
@wwelling Thanks for the feedback! 
Could you elaborate on the context of the sidebar components? In the sidebar.component we have:

```
 constructor(private contextProvider: ContextProviderService, private sidebarService : SidebarService)
    {
        if(contextProvider != null) {
            // when we update the context, we could update the sidebar.
            this.context = contextProvider.context;
            contextProvider.contextObservable.subscribe(x => this.context = x);
     }
```

But maybe I am misunderstanding you. 

In the "helpers" I am not injecting the Authorization/SidebarService which I would have liked to do. But for injecting on a class without decorators, we could 'force' the generation of metadata to support @Inject by using a decorator anyway. But this gave me some trouble (apart from having the decorator which actually did not belong there either)

I also tried this using the ReflectiveInjector but when doing this, the services (which should be singleton) were not acting accordingly. 

I think you might be talking about those two which should be injected in the helper components. Do you have any suggestions on how to deal with this? (One option that I read was that we could excplicitely create a singleton out of the services, but I'm not sure if this will be side-effect free) 

To clarify, what I think you are saying is that in the constructor of the helper it should be like this

`constructor(@Inject(SidebarService) private sidebarService : SidebarService, @Inject(AuthorizationService) private authorization : AuthorizationService, collection : Collection)`

But like this the app crashes at run-time.

ghost commented:
Sorry, I did miss the the constructor of the sidebar component. That looks good. I looked over it again and the sidebar service makes more sense now. I guess the only consideration now is the injection you also mentioned and then the helpers having mostly redundant properties and methods that can be abstract and extend an interface.

As for injection, the helpers could extend a class that is @Injectable that itself injects the service it needs. This class could hold protected properties to reduce the redundancy shared between all the helpers. An interface could make required methods that could be called within the lifecycle methods if need be. I hope this makes sense. Let me know if I am missing more. Thanks.

DylanMeeus commented:
Thanks William, 

I was first thinking along those lines as well, of making it @Injectable but it would seem (to me) that the helpers themselves need not be injected. Though indeed, by making them injectables we could use the normal Angular2 DI. If I see what you mean I could indeed make a 'super' in the style of 

```
@Injectable()
class SidebarHelper{}
```

Which would generate the metadata for us. But then, for DI to work correctly, they will then need to be part of the `server.ts` if I am not mistaken? 

I did look around for other solutions, but it seems like the current DI of Angular2 offers no other solution to this than the one you mentioned, or creating a singleton out of them ourselves (which still requires us to do the DI using the ReflectiveInjector).

ghost commented:
Yeah, for server side they would need to be in server.ts as providers. But for the client side they can be scoped providers. We haven't done this much but can be better to have services provided more granularity where needed. I would avoid manual singleton implementations.

The helpers would definitely benefit from a super class and an interface. The helpers are instantiated continually while navigating around. If possible it would be better to have a singleton that mutates when navigating instead of instantiating a new helper each time.

ghost commented:
As I mentioned these are considerations and not meant to block the merge. The agile process promotes iterative improvements. I would be happy to work on this as well. Although, I am somewhat on the hook for inline editing of items. I plan to start this soon.

DylanMeeus commented:
Thanks William,
I will see how it turns out with scoped providers :-)

DylanMeeus commented:
@wwelling 
I have done as you suggested, the helpers to populate the sidebar are now injected with DI using scoped providers. This has indeed allowed me to inject the AuthorizationService / SidebarService into the Helpers without any issue.

I have also created the (abstract) super class as you suggested. So there is less code-duplication. I do have to pass the SidebarService with the constructor here, because in Typescript (and many other languages), when you have a parent class you will need to call the constructor through the child.

So, if I wanted to Inject in the parent class, it would give issues when calling

```
parentHelper(@Inject(SidebarService) protected sidebarService){ }

childHelper()
{
super(); // not providing the expected number of arguments
}
```

But that was the only "issue" ;-)

ghost commented:
Nice. Do you feel it is an improvement? Thanks for addressing the concerns. 

If anyone else can test and verify, I give a thumbs up on the merge. I will not be able to test until tomorrow. Attending a conference today.

tdonohue commented:
I just mentioned to @DylanMeeus via IRC that I can give this a test run / review today. As long as everything works and looks good, I'll get it merged and we can do any additional cleanup later.

DylanMeeus commented:
Thanks William / Tim.

I think that it is an improvement because now we can just inject the SidebarService and Authorization in the Helpers. But I would have liked to see a better solution to this from Angular2, because now they had to be registered as scoped providers themselves. Though, they are not really meant to be used anywhere else than with their accompanying component. Yet at the moment, this was the only way to deal with it, without passing them via the constructor. It does seem bit of a shame that in order to use the Dependency Injection, another class had to be made Injectable. But, we might be able to change that if Angular changes the way they deal with DI. As you both said, improvements can happen later. ;-)

The parent SidebarHelper was a great suggestion though, it's a lot more `T-DRY` now :-)

tdonohue commented:
Just tested this and it seems to work great.  But, I just saw @DylanMeeus  is still adding commits. Is there more work to be done here, @DylanMeeus?   If so, I can wait to do a final review/merge until it is ready to go.

DylanMeeus commented:
I just finished Tim :-)

tdonohue commented:
Ok, will pull down the latest and give it another quick test/review then. Thanks @DylanMeeus!

DylanMeeus commented:
Good, thanks for all the testing and feedback :)

tdonohue commented:
Tested and it's working great. I also like the new direction of the code. Any further changes/tweaks can be added as additional PRs.
